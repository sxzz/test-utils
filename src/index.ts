export * from './fixture'
export * from './rolldown'
export * from './rollup'
export * from './snapshot'
export * from './string'
