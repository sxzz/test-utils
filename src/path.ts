export function normalizePath(path: string): string {
  return path.replaceAll('\\', '/')
}

export function replacePath(
  source: string,
  search: string,
  replacement: string,
): string {
  return source
    .replaceAll(search, replacement)
    .replaceAll(normalizePath(search), replacement)
}
