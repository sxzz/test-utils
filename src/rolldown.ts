import {
  rolldown,
  type InputOptions,
  type OutputOptions,
  type RolldownOutput,
} from 'rolldown'

type InputPluginOption = NonNullable<InputOptions['plugins']>

export async function rolldownBuild(
  file: string,
  plugins: InputPluginOption = [],
  inputOptions: InputOptions = {},
  outputOptions: OutputOptions = {},
): Promise<{
  chunks: RolldownOutput['output']
  snapshot: string
}> {
  const bundle = await rolldown({
    input: [file],
    treeshake: false,
    onwarn(warning, defaultHandler) {
      if (
        ['UNUSED_EXTERNAL_IMPORT', 'UNRESOLVED_IMPORT'].includes(warning.code!)
      )
        return
      defaultHandler(warning)
    },
    ...inputOptions,
    plugins: [...plugins, ...(inputOptions.plugins || [])],
  })
  const output = await bundle.generate({
    format: 'esm',
    sourcemap: false,
    ...outputOptions,
  })
  const chunks = output.output
  return {
    chunks,
    snapshot: chunks
      .map((file) =>
        file.type === 'chunk'
          ? `// ${file.fileName}\n${file.code}`
          : typeof file.source === 'string'
            ? `// ${file.fileName}\n${file.source}`
            : `// ${file.fileName}\n[BINARY]`,
      )
      .join('\n'),
  }
}
