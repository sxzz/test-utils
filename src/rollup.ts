import {
  rollup,
  type InputOptions,
  type InputPluginOption,
  type OutputOptions,
  type Plugin,
  type RollupOutput,
} from 'rollup'

export async function rollupBuild(
  file: string,
  plugins: InputPluginOption = [],
  inputOptions: InputOptions = {},
  outputOptions: OutputOptions = {},
): Promise<{
  chunks: RollupOutput['output']
  snapshot: string
}> {
  const bundle = await rollup({
    input: [file],
    treeshake: false,
    onwarn(warning, defaultHandler) {
      if (
        ['UNUSED_EXTERNAL_IMPORT', 'UNRESOLVED_IMPORT'].includes(warning.code!)
      )
        return
      defaultHandler(warning)
    },
    ...inputOptions,
    plugins: [plugins, inputOptions.plugins],
  })
  const output = await bundle.generate({
    format: 'esm',
    sourcemap: false,
    ...outputOptions,
  })
  const chunks = output.output
  return {
    chunks,
    snapshot: chunks
      .map((file) =>
        file.type === 'chunk'
          ? file.code
          : typeof file.source === 'string'
            ? file.source
            : file.fileName,
      )
      .join('\n'),
  }
}

export const RollupToStringPlugin = (): Plugin => {
  return {
    name: 'to-string',
    transform: (code) => `export default \`${code.replaceAll('`', '\\`')}\``,
  }
}

export const RollupEscapeNullCharacterPlugin = (): Plugin => {
  return {
    name: 'escape-null-character',
    generateBundle(options, bundle) {
      for (const filename of Object.keys(bundle)) {
        const b = bundle[filename]
        if (b.type !== 'chunk') continue
        if (b.code.includes('\0')) {
          b.code = b.code.replaceAll('\0', '[NULL]')
        }
      }
    },
  }
}
