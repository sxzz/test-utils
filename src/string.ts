export function removeSpaces(s: string): string {
  return s
    .trim()
    .replaceAll(/[\n\r]/g, '')
    .replaceAll(/\s+/g, ' ')
}
