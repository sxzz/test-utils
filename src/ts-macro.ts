import { getLanguagePlugins } from '@ts-macro/language-plugin'
import { proxyCreateProgram } from '@volar/typescript'
import * as ts from 'typescript'
import type { Plugin } from 'ts-macro'

export function createTsMacroProgram(
  rootNames: string[],
  plugins: Plugin[],
  compilerOptions: ts.CompilerOptions = {},
): ts.Program {
  const host = ts.createCompilerHost(compilerOptions)
  const createProgram = proxyCreateProgram(
    ts,
    ts.createProgram,
    (ts, options) => getLanguagePlugins(ts, options.options, { plugins }),
  )

  const program = createProgram({
    options: compilerOptions,
    host,
    rootNames,
  })
  return program
}
