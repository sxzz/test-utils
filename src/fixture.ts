import path from 'node:path'
import process from 'node:process'
import { normalizePath } from '@rollup/pluginutils'
import { glob, type GlobOptions } from 'tinyglobby'
import { describe, test } from 'vitest'

type SkipFn = (testName: string) => boolean | Promise<boolean>
let isSkip: SkipFn | undefined
export function testFixturesSkip(fn: SkipFn): void {
  isSkip = fn
}

export interface FixtureOptions {
  params?: [name: string, values?: any[]][]
  promise?: boolean
  concurrent?: boolean
}

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & FixtureOptions,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: FixtureOptions,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  {
    params,
    promise,
    concurrent,
    ...globOptions
  }: GlobOptions & FixtureOptions = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<FixtureOptions['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          let testFn = test.skipIf(isSkip?.(testName))
          if (concurrent) {
            testFn = testFn.concurrent
          }
          testFn(testName, async ({ expect }) => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(
                currArgs,
                path.resolve(globOptions.cwd || process.cwd(), id),
                code!,
              )
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}

function getName(name: string, value: any) {
  return value !== undefined ? `${name} = ${String(value)}` : name
}
