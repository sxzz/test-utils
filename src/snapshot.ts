import { readFile } from 'node:fs/promises'
import path from 'node:path'
import process from 'node:process'
import { glob } from 'tinyglobby'
import type { OutputFile } from 'esbuild'
import type {
  OutputAsset as RolldownOutputAsset,
  OutputChunk as RolldownOutputChunk,
} from 'rolldown'
import type { OutputAsset, OutputChunk } from 'rollup'
import type { ExpectStatic } from 'vitest'

export function outputToSnapshot(
  chunks: (
    | OutputChunk
    | OutputAsset
    | RolldownOutputChunk
    | RolldownOutputAsset
    | OutputFile
  )[],
): string {
  const cwd = process.cwd()
  return chunks
    .map((file) => {
      let filename: string, content: string
      if ('path' in file) {
        filename = file.path
        content = file.text
      } else {
        filename = file.fileName
        content =
          file.type === 'chunk'
            ? file.code
            : typeof file.source === 'string'
              ? file.source
              : '[BINARY]'
      }
      return `// ${filename.replaceAll('\\', '/')}\n${content.replaceAll(cwd, '[CWD]')}`
    })
    .sort()
    .join('\n')
}

export async function expectFilesSnapshot(
  sourceDir: string,
  snapshotFile: string,
  {
    pattern = '**/*',
    expect,
  }: { pattern?: string; expect?: ExpectStatic } = {},
): Promise<{
  files: string[]
  fileMap: any
  snapshot: string
}> {
  if (!expect) {
    expect = await import('vitest').then((m) => m.expect)
  }

  const cwd = process.cwd()
  const files = (await glob(pattern, { cwd: sourceDir })).sort()
  const fileMap = Object.fromEntries(
    await Promise.all(
      files.map(
        async (filename): Promise<[string, string]> => [
          filename.replaceAll('\\', '/'),
          (
            await readFile(path.resolve(sourceDir, filename), 'utf8')
          ).replaceAll(cwd, '[CWD]'),
        ],
      ),
    ),
  )
  const snapshot = Object.entries(fileMap)
    .map(([filename, contents]) => {
      const ext = path.extname(filename).slice(1)
      return `## ${filename}

\`\`\`${ext}
${contents}
\`\`\``
    })
    .join('\n')
  await expect!(snapshot).toMatchFileSnapshot(snapshotFile)

  return {
    files,
    fileMap,
    snapshot,
  }
}
