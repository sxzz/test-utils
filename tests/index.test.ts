import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'
import { rollupBuild, testFixtures } from '../src'

test('rollupBuild', async () => {
  const { snapshot } = await rollupBuild('tests/fixtures/main.js')
  expect(snapshot).toMatchSnapshot()
})

describe('testFixtures', async () => {
  await testFixtures('tests/fixtures/*.js', (args, id) => {
    expect(args).toEqual({
      'tests/fixtures/main.js': undefined,
    })
    expect(normalizePath(id)).contain('tests/fixtures/main.js')
    return 'ok'
  })
})
