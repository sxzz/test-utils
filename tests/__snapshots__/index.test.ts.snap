// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rollupBuild 1`] = `
"// main.js
const foo = 1;

export { foo };
"
`;

exports[`testFixtures > tests/fixtures/main.js 1`] = `"ok"`;
